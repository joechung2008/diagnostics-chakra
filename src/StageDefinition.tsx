import { Table, Text } from "@chakra-ui/react";
import { useMemo } from "react";

const StageDefinition: React.FC<StageDefinitionProps> = ({
  stageDefinition,
}) => {
  const items = useMemo(
    () =>
      Object.entries(stageDefinition).reduce<KeyValuePair<string[]>[]>(
        (previous, [key, value]) => [...previous, { key, value }],
        []
      ),
    [stageDefinition]
  );

  return (
    <div>
      <Text id="stage-definitions" as="h3" fontSize="lg" fontWeight="semibold">
        Stage Definitions
      </Text>
      <Table.Root aria-labelledby="stage-definitions">
        <Table.Header>
          <Table.Row>
            <Table.ColumnHeader>Key</Table.ColumnHeader>
            <Table.ColumnHeader>Value</Table.ColumnHeader>
          </Table.Row>
        </Table.Header>
        <Table.Body>
          {items.map((item) => (
            <Table.Row key={item.key}>
              <Table.Cell>{item.key}</Table.Cell>
              <Table.Cell>{item.value.join(", ")}</Table.Cell>
            </Table.Row>
          ))}
        </Table.Body>
      </Table.Root>
    </div>
  );
};

export default StageDefinition;
