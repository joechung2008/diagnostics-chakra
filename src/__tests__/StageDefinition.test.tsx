import { describe, it, expect } from "vitest";
import { render } from "@testing-library/react";
import StageDefinition from "../StageDefinition";
import { TestWrapper } from "./test-utils";

describe("StageDefinition", () => {
  it("should render stage definitions heading", () => {
    const stageDefinition = { stage1: ["step1"] };

    const { asFragment } = render(
      <TestWrapper>
        <StageDefinition stageDefinition={stageDefinition} />
      </TestWrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render table with correct headers", () => {
    const stageDefinition = { stage1: ["step1"] };

    const { asFragment } = render(
      <TestWrapper>
        <StageDefinition stageDefinition={stageDefinition} />
      </TestWrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render single stage with single step", () => {
    const stageDefinition = { build: ["compile"] };

    const { asFragment } = render(
      <TestWrapper>
        <StageDefinition stageDefinition={stageDefinition} />
      </TestWrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render stage with multiple steps joined by comma", () => {
    const stageDefinition = { deploy: ["build", "test", "package"] };

    const { asFragment } = render(
      <TestWrapper>
        <StageDefinition stageDefinition={stageDefinition} />
      </TestWrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render multiple stages", () => {
    const stageDefinition = {
      build: ["compile", "lint"],
      test: ["unit", "integration"],
      deploy: ["package", "upload"],
    };

    const { asFragment } = render(
      <TestWrapper>
        <StageDefinition stageDefinition={stageDefinition} />
      </TestWrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should handle empty stage definition", () => {
    const stageDefinition = {};

    const { asFragment } = render(
      <TestWrapper>
        <StageDefinition stageDefinition={stageDefinition} />
      </TestWrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should handle stage with empty steps array", () => {
    const stageDefinition = { emptyStage: [] };

    const { asFragment } = render(
      <TestWrapper>
        <StageDefinition stageDefinition={stageDefinition} />
      </TestWrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it("should render table with correct accessibility attributes", () => {
    const stageDefinition = { stage1: ["step1"] };

    const { asFragment } = render(
      <TestWrapper>
        <StageDefinition stageDefinition={stageDefinition} />
      </TestWrapper>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
