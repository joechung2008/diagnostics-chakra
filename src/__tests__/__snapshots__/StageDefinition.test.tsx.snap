// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StageDefinition > should handle empty stage definition 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div>
    <h3
      class="css-okktu3"
      id="stage-definitions"
    >
      Stage Definitions
    </h3>
    <table
      aria-labelledby="stage-definitions"
      class="chakra-table__root css-a2tetu"
    >
      <thead
        class="chakra-table__header"
      >
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <th
            class="chakra-table__columnHeader css-kim7sg"
          >
            Key
          </th>
          <th
            class="chakra-table__columnHeader css-kim7sg"
          >
            Value
          </th>
        </tr>
      </thead>
      <tbody
        class="chakra-table__body"
      />
    </table>
  </div>
</DocumentFragment>
`;

exports[`StageDefinition > should handle stage with empty steps array 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div>
    <h3
      class="css-okktu3"
      id="stage-definitions"
    >
      Stage Definitions
    </h3>
    <table
      aria-labelledby="stage-definitions"
      class="chakra-table__root css-a2tetu"
    >
      <thead
        class="chakra-table__header"
      >
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <th
            class="chakra-table__columnHeader css-kim7sg"
          >
            Key
          </th>
          <th
            class="chakra-table__columnHeader css-kim7sg"
          >
            Value
          </th>
        </tr>
      </thead>
      <tbody
        class="chakra-table__body"
      >
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <td
            class="chakra-table__cell css-1luypx0"
          >
            emptyStage
          </td>
          <td
            class="chakra-table__cell css-1luypx0"
          />
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`StageDefinition > should render multiple stages 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div>
    <h3
      class="css-okktu3"
      id="stage-definitions"
    >
      Stage Definitions
    </h3>
    <table
      aria-labelledby="stage-definitions"
      class="chakra-table__root css-a2tetu"
    >
      <thead
        class="chakra-table__header"
      >
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <th
            class="chakra-table__columnHeader css-kim7sg"
          >
            Key
          </th>
          <th
            class="chakra-table__columnHeader css-kim7sg"
          >
            Value
          </th>
        </tr>
      </thead>
      <tbody
        class="chakra-table__body"
      >
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <td
            class="chakra-table__cell css-1luypx0"
          >
            build
          </td>
          <td
            class="chakra-table__cell css-1luypx0"
          >
            compile, lint
          </td>
        </tr>
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <td
            class="chakra-table__cell css-1luypx0"
          >
            test
          </td>
          <td
            class="chakra-table__cell css-1luypx0"
          >
            unit, integration
          </td>
        </tr>
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <td
            class="chakra-table__cell css-1luypx0"
          >
            deploy
          </td>
          <td
            class="chakra-table__cell css-1luypx0"
          >
            package, upload
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`StageDefinition > should render single stage with single step 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div>
    <h3
      class="css-okktu3"
      id="stage-definitions"
    >
      Stage Definitions
    </h3>
    <table
      aria-labelledby="stage-definitions"
      class="chakra-table__root css-a2tetu"
    >
      <thead
        class="chakra-table__header"
      >
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <th
            class="chakra-table__columnHeader css-kim7sg"
          >
            Key
          </th>
          <th
            class="chakra-table__columnHeader css-kim7sg"
          >
            Value
          </th>
        </tr>
      </thead>
      <tbody
        class="chakra-table__body"
      >
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <td
            class="chakra-table__cell css-1luypx0"
          >
            build
          </td>
          <td
            class="chakra-table__cell css-1luypx0"
          >
            compile
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`StageDefinition > should render stage definitions heading 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div>
    <h3
      class="css-okktu3"
      id="stage-definitions"
    >
      Stage Definitions
    </h3>
    <table
      aria-labelledby="stage-definitions"
      class="chakra-table__root css-a2tetu"
    >
      <thead
        class="chakra-table__header"
      >
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <th
            class="chakra-table__columnHeader css-kim7sg"
          >
            Key
          </th>
          <th
            class="chakra-table__columnHeader css-kim7sg"
          >
            Value
          </th>
        </tr>
      </thead>
      <tbody
        class="chakra-table__body"
      >
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <td
            class="chakra-table__cell css-1luypx0"
          >
            stage1
          </td>
          <td
            class="chakra-table__cell css-1luypx0"
          >
            step1
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`StageDefinition > should render stage with multiple steps joined by comma 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div>
    <h3
      class="css-okktu3"
      id="stage-definitions"
    >
      Stage Definitions
    </h3>
    <table
      aria-labelledby="stage-definitions"
      class="chakra-table__root css-a2tetu"
    >
      <thead
        class="chakra-table__header"
      >
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <th
            class="chakra-table__columnHeader css-kim7sg"
          >
            Key
          </th>
          <th
            class="chakra-table__columnHeader css-kim7sg"
          >
            Value
          </th>
        </tr>
      </thead>
      <tbody
        class="chakra-table__body"
      >
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <td
            class="chakra-table__cell css-1luypx0"
          >
            deploy
          </td>
          <td
            class="chakra-table__cell css-1luypx0"
          >
            build, test, package
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`StageDefinition > should render table with correct accessibility attributes 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div>
    <h3
      class="css-okktu3"
      id="stage-definitions"
    >
      Stage Definitions
    </h3>
    <table
      aria-labelledby="stage-definitions"
      class="chakra-table__root css-a2tetu"
    >
      <thead
        class="chakra-table__header"
      >
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <th
            class="chakra-table__columnHeader css-kim7sg"
          >
            Key
          </th>
          <th
            class="chakra-table__columnHeader css-kim7sg"
          >
            Value
          </th>
        </tr>
      </thead>
      <tbody
        class="chakra-table__body"
      >
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <td
            class="chakra-table__cell css-1luypx0"
          >
            stage1
          </td>
          <td
            class="chakra-table__cell css-1luypx0"
          >
            step1
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`StageDefinition > should render table with correct headers 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div>
    <h3
      class="css-okktu3"
      id="stage-definitions"
    >
      Stage Definitions
    </h3>
    <table
      aria-labelledby="stage-definitions"
      class="chakra-table__root css-a2tetu"
    >
      <thead
        class="chakra-table__header"
      >
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <th
            class="chakra-table__columnHeader css-kim7sg"
          >
            Key
          </th>
          <th
            class="chakra-table__columnHeader css-kim7sg"
          >
            Value
          </th>
        </tr>
      </thead>
      <tbody
        class="chakra-table__body"
      >
        <tr
          class="chakra-table__row css-11ot24d"
        >
          <td
            class="chakra-table__cell css-1luypx0"
          >
            stage1
          </td>
          <td
            class="chakra-table__cell css-1luypx0"
          >
            step1
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
